[manifold] Add new port

JeffreyWardman opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Commit:
manifold port

Commit:
add manifold to version database

JeffreyWardman commented:
@microsoft-github-policy-service agree

Commit:
ignore warnings

Commit:
update version

JeffreyWardman commented:
I was able to build locally. manifold builds files into `lib/cmake` which raises warnings, which seem to have caused the CI/CD to fail. Once adding the ignore warnings, I get the current errors only in CI/CD.

Initial error due to warnings: https://dev.azure.com/vcpkg/public/_build/results?buildId=109648&view=logs&j=878666d5-db33-5b27-9e7d-b0c7ee352005&t=69176ca8-c93b-5311-0f74-6cb12a2b3f92&l=8209

After adding the ignore fixes: https://dev.azure.com/vcpkg/public/_build/results?buildId=109649&view=logs&j=878666d5-db33-5b27-9e7d-b0c7ee352005&t=69176ca8-c93b-5311-0f74-6cb12a2b3f92&l=8184

This is my first commit to vcpkg. It'd be great if I can get advice on how to handle these kinds of warnings for it to be merged.

BillyONeal reviewed (dismissed):
> This is my first commit to vcpkg. It'd be great if I can get advice on how to handle these kinds of warnings for it to be merged.

Usually these kinds of problems require patching upstream's build system, or moving around bits in portfile.cmake.

Do you have specific questions about the warnings? They seem pretty explicit in what the problem is to me, but then again, I wrote most of the messages 😅

Repeating the errors here so that the logs aren't lost when the build gets purged in a few days:

> D:\a\_work\1\s\ports\manifold\portfile.cmake: warning: ${CURRENT_PACKAGES_DIR}/debug/include should not exist. To suppress this message, add set(VCPKG_POLICY_ALLOW_DEBUG_INCLUDE enabled)
> note: If this directory was created by a build system that does not allow installing headers in debug to be disabled, delete the duplicate directory with file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
> D:\a\_work\1\s\ports\manifold\portfile.cmake: warning: This port installs the following CMake files in places CMake files are not expected. CMake files should be installed in ${CURRENT_PACKAGES_DIR}/share/${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
> D:\p\manifold_x86-windows: note: the files are relative to ${CURRENT_PACKAGES_DIR} here
> note: lib/cmake/manifold/manifoldConfig.cmake
> note: lib/cmake/manifold/manifoldConfigVersion.cmake
> note: lib/cmake/manifold/manifoldTargets-release.cmake
> note: lib/cmake/manifold/manifoldTargets.cmake
> note: debug/lib/cmake/manifold/manifoldConfig.cmake
> note: debug/lib/cmake/manifold/manifoldConfigVersion.cmake
> note: debug/lib/cmake/manifold/manifoldTargets-debug.cmake
> note: debug/lib/cmake/manifold/manifoldTargets.cmake
> D:\a\_work\1\s\ports\manifold\portfile.cmake: warning: This port creates ${CURRENT_PACKAGES_DIR}/lib/cmake and/or ${CURRENT_PACKAGES_DIR}/debug/lib/cmake, which should be merged and moved to ${CURRENT_PACKAGES_DIR}/share/${PORT}/cmake. Please use the helper function vcpkg_cmake_config_fixup() from the port vcpkg-cmake-config. To suppress this message, add set(VCPKG_POLICY_SKIP_LIB_CMAKE_MERGE_CHECK enabled)
> D:\a\_work\1\s\ports\manifold\portfile.cmake: warning: Found 3 post-build check problem(s). These are usually caused by bugs in portfile.cmake or the upstream build system. Please correct these before submitting this port to the curated registry.
> error: building manifold:x86-windows failed with: POST_BUILD_CHECKS_FAILED

JeffreyWardman commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+add set(VCPKG_POLICY_ALLOW_DEBUG_INCLUDE enabled)
+set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
+add set(VCPKG_POLICY_SKIP_LIB_CMAKE_MERGE_CHECK enabled)
```
What's the best approach to do that? Create a patch that moves its location and the paths within the target cmake files?

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO elalish/manifold
+    REF v3.0.0
```
```suggestion
    REF v${VERSION}
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+    SHA512 881d3b0e3ff03794ce66b09c4a7be675e5dcd5d5b269d62ad5c5de177e76a01460f6f0fb55a2973a92abda3bf32b8a08bafdff5c0b379ae095d9806eb5669022
+)
+
+# Configure the project
```
```suggestion
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+
+# Configure the project
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
```
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+    SOURCE_PATH ${SOURCE_PATH}
+    OPTIONS
+        -DMANIFOLD_DEBUG=OFF
+        -DMANIFOLD_TEST=OFF  # Disable tests to avoid overhead
+        -DMANIFOLD_CROSS_SECTION=ON  # Enable cross-section support
+        -DMANIFOLD_CBIND=ON  # Enable C bindings
+        -DMANIFOLD_PYBIND=OFF  # Enable Python bindings
+        -DMANIFOLD_JSBIND=OFF  # Disable JS bindings
```
```suggestion
        -DMANIFOLD_TEST=OFF
        -DMANIFOLD_CROSS_SECTION=ON
        -DMANIFOLD_CBIND=ON
        -DMANIFOLD_PYBIND=OFF
        -DMANIFOLD_JSBIND=OFF
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+        -DMANIFOLD_CBIND=ON  # Enable C bindings
+        -DMANIFOLD_PYBIND=OFF  # Enable Python bindings
+        -DMANIFOLD_JSBIND=OFF  # Disable JS bindings
+    OPTIONS_RELEASE
+        -DMANIFOLD_DEBUG=OFF
```
```suggestion
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+        -DMANIFOLD_DEBUG=ON
+)
+
+# Build and install
+vcpkg_cmake_build()
```
```suggestion
```

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,36 @@
+
+# Build and install
+vcpkg_cmake_build()
+vcpkg_cmake_install()
```
This should resolve the issue with lib/cmake. Please use the actual path.
```suggestion
vcpkg_cmake_install()
vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/manifold)
```

dg0yt commented on ports/manifold/usage:
```diff
@@ -0,0 +1,4 @@
+The package manifold provides CMake targets:
+
+    find_package(manifold CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE manifold::manifold)
```
Formatting as printed by the heuristics:
```suggestion
manifold provides CMake targets:

  find_package(manifold CONFIG REQUIRED)
  target_link_libraries(main PRIVATE manifold::manifold)
```
If the heuristics has the same output, a usage file is redundant.

Commit:
address comments

Commit:
update manifold version git tree

JeffreyWardman commented:
Waiting for CI/CD pipeline to complete but I expect that I need to add:
```
  "supports": "((x64 & (windows | osx | linux)) | (x86 & windows) | (arm64 & osx)) & !uwp & !static",
```

JeffreyWardman commented:
So, it supports:
`x86_windows`, `x64_windows`, `x64_windows_static_md`, `x64_uwp`, `arm64_windows`, `arm64_windows_static_md`, `arm64_uwp`, `x64_osx`, `arm64_osx`, `x64_linux`.

It does not support `x64_windows_static`, `arm_neon_android`, `x64_android`, `arm64_android`.

How can I represent this? What do I extend onto `"supports": "!android" to include `x64_windows_static` only?

dg0yt commented:
The failures for android and x64-windows-static are due to `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` in triplets with static crt linkage. This shall not be reflected in supports but as expected fail in script/ci.baseline.txt. (User can overcome the guard via per-port customization.)

Commit:
remove check for dynamic lib

Commit:
update manifold git tree

Commit:
add manifold debug option

Commit:
update manifold git tree

Commit:
remove debug option from manifold port

Commit:
update manifold version git tree

JeffreyWardman commented:
Thanks so much for your review and help so far @dg0yt and @BillyONeal. I've added the requested changes and removed the (unneeded) dynamic lib check.

All CI/CD pipelines pass if built in release mode but there is an error when compiling for `arm-neon-android` in debug mode only (an issue on `manifold`'s end). I've created a PR for the fix: https://github.com/elalish/manifold/pull/1077. As this is their 3.0 release, it'd be good to push up the current version.

I've reverted enabling debug builds with the idea that the initial port gets pushed with v3.0.0 and a ticket is created for enabling debug builds that's dependent on the next patch.

Please let me know if an alternative is preferred.

JeffreyWardman commented:
Thanks so much for your review and help so far @dg0yt and @BillyONeal. I've added the requested changes and removed the (unneeded) dynamic lib check.

All CI/CD pipelines pass if built in release mode but there is an error when compiling for `arm-neon-android` in debug mode only (an issue on `manifold`'s end). I've created a PR for the fix: https://github.com/elalish/manifold/pull/1077. As this is their 3.0 release, it'd be good to push up the current version.

I've reverted enabling debug builds with the idea that the initial port gets pushed with v3.0.0 and a ticket is created for enabling debug builds that's dependent on the next patch.

Please let me know if an alternative is preferred.

dg0yt commented on ports/manifold/usage:
```diff
@@ -0,0 +1,4 @@
+manifold provides CMake targets:
+
+  find_package(manifold CONFIG REQUIRED)
+  target_link_libraries(main PRIVATE manifold::manifold)
```
We somehow lost a linebreak at the end of this file.

dg0yt commented on ports/manifold/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DMANIFOLD_DEBUG=OFF
```
AFAIU this doesn't control what we call "debug builds", but is an additional "debug tracing/timing" capability. Maybe it can remain turned off. That is the default anyways.
(Relevant debug build compile options come with the toolchain. For MSVC, the most important thing about "debug builds" is linking to the debug runtime.)

dg0yt replied:
That line doesn't change anything. So there is no need to keep it.

Commit:
remove unused lines in manifold port

Commit:
update manifold version git tree

JeffreyWardman commented:
Hi all, is there anything blocking this PR from being merged now?

pca006132 commented:
@dg0yt ping

dg0yt reviewed (approved):
LGTM but somebody else must approve.
@FrankXie05?

FrankXie05 commented:
The usage has been tested successfully locally on `x64-windows`.
